### DISCUSSION TITLE
I don't know if target is being instrumented correctly

### DISCUSSION THREAD

**Discussion by andrecoco** [AFLplusplus/AFLplusplus#1629] [Q&A]
Hi, I was fuzzing a target, and the fuzzer was taking too long to find new branches, so I check the code to see if the partial instrumentation was working as intended. And I got confused when looking at the assembly, because I had instrumented some other binaries before, and they all had '__afl_area_ptr' in every new branch, but this target only has **on the beginning of each function**.

I tried with and without partial instrumentation, as well as compiling with afl-clang-fast and afl-clang-lto, the result was the same in all cases.
Some pictures of the assembly of this target below as example.
![Weird1](https://user-images.githubusercontent.com/16878985/215151915-ffd74745-7542-4edb-b6c3-a16c2763be74.png)
![Weird2](https://user-images.githubusercontent.com/16878985/215151928-3aab10df-1f06-4437-93e9-85462eb0b569.png)

Some pictures of the assembly of other targets I fuzzed before.
![looksNormal1](https://user-images.githubusercontent.com/16878985/215151939-0b7d78ce-f4bf-4e64-8475-809609e48809.png)
![looksNormal2](https://user-images.githubusercontent.com/16878985/215151960-25ad4315-ccd0-4968-bdb2-e1b38a10d5da.png)

I'm not very familiar with how afl does its instrumentation, so I can't be sure that something is wrong, but I found it weird.

I'm building the project with meson, I basically exported LLVM_CONFIG, and added the CC and CXX flags before each meson command.

```
export LLVM_CONFIG="llvm-config"
CC=afl-clang-lto CXX=afl-clang-lto++ 
```
Thanks in advance

**Comment by vanhauser-thc** [comment]
with llvm version are you using? I recommend 14, there are bugs in previous ones

**Reply by vanhauser-thc** [reply]
also this is very wrong:
CC=afl-clang-fast CXX=CC=afl-clang-lto++ 
either you use afl-cang-fast and afl-clang-fast++ or afl-clang-lto and afl-clang-lto++. you cannot mix

**Reply by andrecoco** [reply]
sorry, that was a typo in the last build I made, but all the screenshots are from builds using
CC=afl-clang-lto CXX=afl-clang-lto++ 
or 
CC=afl-clang-fast CXX=afl-clang-fast++

**Reply by andrecoco** [reply]
llvm-config --version returns "15.0.7"

**Reply by vanhauser-thc** [reply]
did you use any other compile options? like AFL_LLVM_LAF_ALL, etc. ?

can you also please try to downgrade to 14 and recompile + install afl++?

**Reply by andrecoco** [reply]
`did you use any other compile options? like AFL_LLVM_LAF_ALL, etc. ?`
In one of the builds yes, actually I was first looking at the assembly to see the differences between the binary compiled with LLVM_LAF_ALL and the "normal" one, and later  I realized the missing "__afl_area_ptr". (and the binaries seemed identical, I don't know if LAF was working as well).

But the screenshot are from builds without the LAF.

`can you also please try to downgrade to 14 and recompile + install afl++?`
I downgraded, reinstalled the AFL and the binary now looks better, although some jumps still don't have the __afl_area_ptr. 

This part now has a lot of __afl_area_ptr:
![Normal_llvm14](https://user-images.githubusercontent.com/16878985/215169851-fcd97ea3-0977-4f1a-8e3c-0e1a99754732.png)

But this other function only has a few (but maybe that's normal behavior?) 
![Noafl_llvm14](https://user-images.githubusercontent.com/16878985/215169902-f044e98f-db81-468a-a43c-b2f041b7d1af.png)

**Reply by vanhauser-thc** [reply]
the last one you doubt is all fine. where a block has two previous blocks you see the movxd rax, cs:dword_xxxx ... that is the coverage. r14 is the coverage map.

so it looks like llvm was able to fuck things up in version 15 again. great.

**Reply by vanhauser-thc** [reply]
is your target public? I just installed llvm 15.0.6 but I see so far no notable difference in the number of instrumentations

**Reply by andrecoco** [reply]
Yes, it's mpv-player (mpv-player/mpv on github). And I'm using the 'mpv-build' (mpv-player/mpv-build) to help compile all the dependency libraries. 
I'm putting the AFL flags on the "/scripts/mpv-config" and "scripts/mpv-build".  The screenshots are from some of the demux functions.
